### PR TITLE
Implement a new overland flow formulation with channel width

### PR DESCRIPTION
**Pull request by kenadigwaymire**
Added reading channel width X and Y files.   The files are not yet used, this change is in preparation for using in a pending  overland flow formulation change that includes channel width as a parameter. Test added to ensure that we could read both constant channel width data and channel width data from PFBs.

### CODE REVIEW

File: pf-keys/definitions/geom.yaml [hunk 1]
```diff
@@ -1480,6 +1488,151 @@ TopoSlopesY:
         domains:
           DoubleValue:
 
+# -----------------------------------------------------------------------------
+# ChannelWidth
+# -----------------------------------------------------------------------------
+# ChannelWidth missing from manual and grep
```
Comment (smithsg84) [inline pf-keys/definitions/geom.yaml:1494]: The users manual should be updated with the new keys.

File: pf-keys/definitions/geom.yaml [hunk 2]
```diff
@@ -1480,6 +1488,151 @@ TopoSlopesY:
...
+        type: ChildrenHandler
+        class_name: ChannelWidthXItem
+        location: ./Geom
+
+  # in manual, need to consolidate duplicates (manual also includes NetCDF file capability, which is included here)
+  Type:
+    help: >
+      [Type: string] This key specifies which method is to be used to assign channel widths. The choices currently available are
+      Constant which indicates that a constant is to be assigned to all grid cells within a geometry and PFBFile which
+      indicates that all values are read in from a distributed, grid-based ParFlow binary file. If NetCDF is used, NCFile can be
```
Comment (smithsg84) [inline pf-keys/definitions/geom.yaml:1535]: Is NetCDF reading slopes here?  Is this a cut-paste error?

File: pf-keys/definitions/geom.yaml [hunk 3]
```diff
@@ -1480,6 +1488,151 @@ TopoSlopesY:
         domains:
           DoubleValue:
 
+# -----------------------------------------------------------------------------
+# ChannelWidth
+# -----------------------------------------------------------------------------
+# ChannelWidth missing from manual and grep
+
+# TODO: Change the documentation to match with channel width, also add handlers if needed? (Not sure if they are needed)
```
Comment (smithsg84) [inline pf-keys/definitions/geom.yaml:1496]: Not sure what first part of this TODO is saying.

File: pfsimulator/parflow_lib/solver_richards.c [hunk 4]
```diff
@@ -611,6 +612,26 @@ SetupRichards(PFModule * this_module)
...
+  {
+    strcpy(file_postfix, "wc_x");
+    WritePFBinary(file_prefix, file_postfix,
+                  ProblemDataChannelWidthX(problem_data));
+
+    strcpy(file_postfix, "wc_y");
+    WritePFBinary(file_prefix, file_postfix,
+                  ProblemDataChannelWidthY(problem_data));
+
+    static const char* slope_filenames[] = {
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/solver_richards.c:626]: Nitpick but these are not "slope_" filenames.   I'm fine with just using "filenames" here to reduce the "copy/paste forget to change name" errors.   A local "filenames" array should be reasonably clear.

File: pfsimulator/parflow_lib/problem_wc_y.c [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+  Vector *wcy_values; 
+} Type1;                       /* .pfb file */
+
+typedef struct {
+  char  *filename;
+  Vector *wcy_values;
+} Type2;                       /* .nc file */
+
+/*
+* YChannelWidth
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/problem_wc_y.c:66]: Methods should be documented using Doxygen format. See

https://github.com/parflow/parflow/blob/master/CONTRIBUTING.md#source-code-documentation

grgeom_octree.h has Doxygen comments as examples.

File: test/python/channelwidth.py [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+wc_test.run(working_directory=base)
+
+#-----------------------------------------------------------------------------
+# Check if values are as expected
+#-----------------------------------------------------------------------------
+
+data_wcx = read_pfb(os.path.join(base, "channel_width_example.out.wc_x.pfb"))
+data_wcy = read_pfb(os.path.join(base, "channel_width_example.out.wc_y.pfb"))
+
+if not np.array_equal(data_wcx, ones_array) or not np.array_equal(data_wcy, ones_array):
```
Comment (smithsg84) [inline test/python/channelwidth.py:401]: This test in not checking very much.  Should check simulation results.

**Review by smithsg84 — CHANGES_REQUESTED** [review]
This PR was not marked as DRAFT but is this done?   I don't see the width is integrated with overland flow.

File: pf-keys/definitions/geom.yaml (hunk 1 above)
Comment (gartavanis) [reply pf-keys/definitions/geom.yaml:1500]: We will generate the manual again once this PR gets merged to main. I think that these keys will appear automatically since they are included in the yaml file?

File: pf-keys/definitions/geom.yaml (hunk 3 above)
Comment (gartavanis) [reply pf-keys/definitions/geom.yaml:1502]: Handlers were also added, this comment should have been removed.

File: pf-keys/definitions/geom.yaml (hunk 2 above)
Comment (gartavanis) [reply pf-keys/definitions/geom.yaml:1541]: Looks like cut-paste error from TopoSlopes.

**Comment by gartavanis** [comment]
> This PR was not marked as DRAFT but is this done? I don't see the width is integrated with overland flow.

File: pf-keys/definitions/geom.yaml (hunk 1 above)
Comment (smithsg84) [reply pf-keys/definitions/geom.yaml:1500]: The documentation RST files are not automatically generated from the YAML documentation.

File: test/python/channelwidth.py (hunk 6 above)
Comment (gartavanis) [reply test/python/channelwidth.py:405]: This PR simply adds the capability to read ChannelWidthX and ChannelWidthY inputs. It is not yet integrated with overland flow. @reedmaxwell suggested breaking this into two separate PRs, because the new overland flow formulation will need different boundary conditions as well and this might take longer to develop. So this test just checks that ParFlow reads correctly the inputs. It can be modified once the integration with overland flow happens, to check against simulation results.

**Review by smithsg84 — APPROVED** [review]
Thanks for updates.
